python/activator/activator.py

```diff
@@ -195,6 +225,105 @@ def create_app():
+        _log.exception(e)
+
+    # Initialize schema registry for fan out
+    fan_out_schema_registry_conf = {'url': fan_out_schema_registry_url}
+    fan_out_schema_registry_client = SchemaRegistryClient(fan_out_schema_registry_conf)
+
+    fan_out_avro_deserializer = AvroDeserializer(schema_registry_client=fan_out_schema_registry_client,
+                                                 from_dict=dict_to_fanned_out_visit)
+    fan_out_consumer_conf = {
+        "bootstrap.servers": fan_out_kafka_cluster,
+        "group.id": fan_out_kafka_group_id,
+        "auto.offset.reset": fan_out_kafka_topic_offset,
+        "sasl.mechanism": fan_out_kafka_sasl_mechanism,
+        "security.protocol": fan_out_kafka_security_protocol,
+        "sasl.username": fan_out_kafka_sasl_username,
+        "sasl.password": fan_out_kafka_sasl_password,
+        'enable.auto.commit': False
+    }
+
+    _log.info("starting fan out consumer")
+    fan_out_consumer = kafka.Consumer(fan_out_consumer_conf, logger=_log)
+    fan_out_consumer.subscribe([fan_out_kafka_topic])
+    fan_out_listen_start_time = time.time()
+
+    try:
+        while time.time() - fan_out_listen_start_time < fanned_out_msg_listen_timeout:
+
+            fan_out_message = fan_out_consumer.poll(timeout=5)
+            if fan_out_message is None:
+                continue
+            if fan_out_message.error():
+                _log.warning("Fanned out consumer error: %s", fan_out_message.error())
+            else:
+                deserialized_fan_out_visit = fan_out_avro_deserializer(fan_out_message.value(),
+                                                                       SerializationContext(
+                                                                       fan_out_message.topic(),
+                                                                       MessageField.VALUE))
+                _log.info("Unpacked message as %r.", deserialized_fan_out_visit)
+
+                # Calculate time to load knative and receive message based on timestamp in Kafka message
+                _log.debug("Message timestamp %r", fan_out_message.timestamp())
+                fan_out_kafka_msg_timestamp = fan_out_message.timestamp()
+                fan_out_to_prompt_time = int(time.time() * 1000) - fan_out_kafka_msg_timestamp[1]
+                _log.debug("Seconds since fan out message delivered %r", fan_out_to_prompt_time/1000)
+
+                # Commit message and close client
+                fan_out_consumer.commit(message=fan_out_message, asynchronous=False)
+                fan_out_consumer.close()
+
+                try:
+                    # Process fan out visit
+                    process_visit(deserialized_fan_out_visit)
+                    _log.info("Processing completed for %s", socket.gethostname())
+                    break
+                except Exception as e:
+                    _log.critical("Process visit failed; aborting.")
+                    _log.exception(e)
```

Comment:
An exception will cause this to loop (and likely fail over the closed consumer). Possible fix:
```suggestion
                try:
                    # Process fan out visit
                    process_visit(deserialized_fan_out_visit)
                except Exception as e:
                    _log.critical("Process visit failed; aborting.")
                    _log.exception(e)
                finally:
                    _log.info("Processing completed for %s", socket.gethostname())
                    break
```